lib/core/text/u16_string.nit

```diff
@@ -0,0 +1,125 @@
+
+`{
+	#include <unicode/utypes.h>
+	#include <unicode/ustring.h>
+`}
+
+# UTF-16 encoded string
+class U16String
+	super Finalizable
+
+	# Pointer to a `UChar *` string
+	var uchar_string : UCharString is noinit
+
+	# Number of characters in `uchar_string`
+	var length = 0
+
+	# Number of bytes in `uchar_string`
+	var bytes = 0
+
+	# Returns a converted `UCharString` from a `String`
+	init (source : String)
+	do
+		var csource = source.to_cstring
+		var csource_length = source.to_cstring.cstring_length
+
+		uchar_string = new UCharString.nul
+		var required_length = uchar_string.from_cstring(0, csource, csource_length)
+
+		uchar_string = new UCharString.empty(required_length + 1)
+		uchar_string.from_cstring(required_length + 1, csource, csource_length)
+
+		length = csource.to_s.length
+		bytes = csource_length
+	end
+
+	# Returns an empty `U16String` of length `len`
+	init empty (len : Int) do
+		uchar_string = new UCharString.empty(len)
+		length = len - 1
+	end
+
+	# Returns a null `U16String`
+	init nul do uchar_string = new UCharString.nul
+
+	# Get char at `index`
+	fun [](index: Int): Char do return uchar_string[index]
+
+	# Set char to `item` at `index`
+	fun []=(index: Int, item: Char) do uchar_string[index] = item
+
+	# Converts the `U16String` to a `CString`
+	fun to_cstring : CString
+	do
+		var cself = new CString.nul
+		var required_length = uchar_string.to_cstring(cself, 0, length)
+
+		cself = new CString(required_length + 1)
+		uchar_string.to_cstring(cself, required_length + 1, length)
+
+		return cself
+	end
+
+	redef fun to_s : String do return to_cstring.to_s
+	redef fun finalize do uchar_string.free
+end
+
+# ICU string `UChar *` which are UTF-16 strings
+extern class UCharString `{ UChar *`}
+
+	# Returns an empty `UCharString` of length `length`
+	new empty (length : Int) `{ return (UChar *)malloc(sizeof(UChar) * length); `}
+
+	# Returns a `NULL` `UCharString`
+	new nul `{ return NULL; `}
+
+	# Converts a `CString` to a `UCharString` and returns the required length (without the termination character) of said `UCharString`
+	fun from_cstring(dest_length : Int, source : CString, source_length : Int) : Int `{
+		UErrorCode error = U_ZERO_ERROR;
+		int32_t res;
+		u_strFromUTF8(self, dest_length, &res, source, source_length, &error);
+		return res;
+	`}
+
+	# Converts `self` to a `CString` and returns the required length (without the termination character) of said `CString`
+	fun to_cstring(dest : CString, dest_length : Int, source_length : Int) : Int `{
+		UErrorCode error = U_ZERO_ERROR;
+		int32_t res;
+		u_strToUTF8(dest, dest_length, &res, self, source_length, &error);
+		return res;
+	`}
+
+	# Get char at `index`
+	fun [](index: Int): Char `{ return self[index]; `}
+
+	# Set char to `item` at `index`
+	fun []=(index: Int, item: Char) `{ self[index] = item; `}
+end
+
+redef class String
+	# Returns a UTF-16 encoded version of `self`
+	fun to_ustring : U16String do return new U16String(self)
```

Comment:
This is probably me nitpicking, but I'd think the method could be named to `to_u16string` since the class returned is `U16String`, and because the standard `String` is already Unicode-compliant for a bunch of cases